core/src/test/java/org/elasticsearch/gateway/GatewayMetaStateTests.java

```diff
@@ -245,4 +259,193 @@ public void testWriteClosedIndex() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        MetaDataUpgrader metaDataUpgrader = new MetaDataUpgrader(
+            Collections.singletonList(customs -> {
+                List<MetaData.Custom> customList = new ArrayList<>();
+                for (MetaData.Custom custom : customs) {
+                    if (custom.type().equals(CustomMetaData1.TYPE)) {
+                        customList.add(new CustomMetaData1("modified_data1"));
+                    } else {
+                        customList.add(custom);
+                    }
+                }
+                return customList;
+            })
+        );
+
+        MetaData upgrade = GatewayMetaState.upgradeMetaData(metaData, new MockMetaDataIndexUpgradeService(false), metaDataUpgrader);
+        assertTrue(upgrade != metaData);
+        assertFalse(MetaData.isGlobalStateEquals(upgrade, metaData));
+        assertNotNull(upgrade.custom(CustomMetaData1.TYPE));
+        assertThat(((TestCustomMetaData) upgrade.custom(CustomMetaData1.TYPE)).getData(), equalTo("modified_data1"));
+        for (IndexMetaData indexMetaData : upgrade) {
+            assertTrue(metaData.hasIndexMetaData(indexMetaData));
+        }
+    }
+
+    public void testNoMetaDataUpgrade() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        MetaDataUpgrader metaDataUpgrader = new MetaDataUpgrader(Collections.emptyList());
+        MetaData upgrade = GatewayMetaState.upgradeMetaData(metaData, new MockMetaDataIndexUpgradeService(false), metaDataUpgrader);
+        assertTrue(upgrade == metaData);
+        assertTrue(MetaData.isGlobalStateEquals(upgrade, metaData));
+        for (IndexMetaData indexMetaData : upgrade) {
+            assertTrue(metaData.hasIndexMetaData(indexMetaData));
+        }
+    }
+
+    public void testCustomMetaDataValidation() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        MetaDataUpgrader metaDataUpgrader = new MetaDataUpgrader(Collections.singletonList(
+            customs -> {
+                throw new IllegalStateException("custom meta data too old");
+            }
+        ));
+        try {
+            GatewayMetaState.upgradeMetaData(metaData, new MockMetaDataIndexUpgradeService(false), metaDataUpgrader);
+        } catch (IllegalStateException e) {
+            assertThat(e.getMessage(), equalTo("custom meta data too old"));
+        }
+    }
+
+    public void testMultipleCustomMetaDataUpgrade() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        MetaDataUpgrader metaDataUpgrader = new MetaDataUpgrader(
+            Arrays.asList(
+                customs -> {
+                    List<MetaData.Custom> customList = new ArrayList<>();
+                    for (MetaData.Custom custom : customs) {
+                        if (custom.type().equals(CustomMetaData1.TYPE)) {
```

Comment:
I notice this pattern in every implementation. Perhaps this should be a Map instead of Collection (keyed by the custom type name)? Then the map can be copied, and keys replaced, removed, or added easily, without needing to have logic for the other custom metadata that the plugin does not care about.